core/primitives/src/offchain.rs

```diff
@@ -430,6 +475,147 @@ impl<T: Externalities + ?Sized> Externalities for Box<T> {
 		(&mut **self).http_response_read_body(request_id, buffer, deadline)
 	}
 }
+/// An `OffchainExternalities` implementation with limited capabilities.
+pub struct LimitedExternalities<T> {
+	capabilities: Capabilities,
+	externalities: T,
+}
+
+impl<T> LimitedExternalities<T> {
+	/// Create new externalities limited to given `capabilities`.
+	pub fn new(capabilities: Capabilities, externalities: T) -> Self {
+		Self {
+			capabilities,
+			externalities,
+		}
+	}
+
+	/// Check if given capability is allowed.
+	///
+	/// Returns `Ok` in case it is, `Err` otherwise.
+	fn check(&self, capability: Capability) -> Result<(), ()> {
+		if self.capabilities.has(capability) {
+			Ok(())
+		} else {
+			#[cfg(feature = "log")]
+			log::warn!("Accessing a forbidden API. No: {:?} capability.", capability);
+			Err(())
+		}
+	}
+}
+
+impl<T: Externalities> Externalities for LimitedExternalities<T> {
+	fn is_validator(&self) -> bool {
+		if self.check(Capability::Keystore).is_ok() {
```

Comment:
`is_validator` does not accesses the keystore. Maybe we just need a `General` capability or something similar.

Reply:
To be honest I question the existence of this API at all:
1. The only reason I can see to check if someone is a validator is to avoid running the code that would fail at signing stuff (i.e. requires keystore anyway). I'd rather test if keystore public keys are on the session list instead - the fact that someone is configured as a validator doesn't really mean that they are validating.
2. Offchain workers are disabled for non-validators by default, if someone enables them from CLI this check would unnecessarily prevent them from doing so. Note this check is by no means a security measure, since everyone can just take the offchain worker code and run it without the check anyway.

I'd rather have a way to filter out offchain worker modules one wants to run from CLI than introducing such checks.

Reply:
We introduced the code, so that nodes with a key that are not sending im-online messages. If only validators are running offchain-code. Then there is no reason at all for this code. So, we should remove the function completely. Maybe keep a stub to not break kusama, but always return true and remove the usage in im-online.

Reply:
If we want to remove it we should do it in a separate PR - it's not related. I'm fine with leaving it, since it's already here, but don't think there is enough reasons to introduce a separate capability for it - it would be a super-set of `Keystore` anyway.

Reply:
Yeah